File: apple/RecKit/Sources/RecKit/Jobs/RecordingPlayer.swift
import AVFoundation
import Foundation
import ReclyCore

/// docs/08 "결과 파일": which of the files beside `meta.json` the detail plays back, and in what
/// order. A pure choice over `meta`, so it can be checked without a disk or a player — the shell
/// only hands it the recording's directory and a way to ask whether a file is there.
public enum RecordingPlaylist {
    /// The local audio of one recording, in play order: what to play, and how long each part is.
    /// The durations are the ones `meta.json` recorded, which is what the elapsed clock counts in —
    /// asking AVFoundation would mean loading every file before the first one could start.
    public struct Selection: Equatable, Sendable {
        public let urls: [URL]
        public let durations: [Double]

        public static let empty = Selection(urls: [], durations: [])

        public var isEmpty: Bool { urls.isEmpty }
        public var totalSec: Double { durations.reduce(0, +) }
    }

    /// The one track a person means by "play this": the mix if the recording has one — a meeting's
    /// mic and system audio already summed — and otherwise the single `mono` track a memo is.
    /// `mic`/`sys` are the mix's own ingredients and are never played on their own here.
    public static func playedTrack(tracks: [Track]) -> Track {
        tracks.contains(Track.mix) ? Track.mix : Track.mono
    }

    /// A part whose file is not on this device is dropped rather than played as silence: what the
    /// retention sweep took leaves a playlist with a gap in it (docs/03 ADR-017), which the detail
    /// either fills from Drive ([fetchesFromDrive]) or says in words.
    public static func select(
        tracks: [Track],
        parts: [Part_],
        dir: URL,
        exists: (URL) -> Bool
    ) -> Selection {
        let track = playedTrack(tracks: tracks)
        let kept = parts
            .filter { $0.track == track }
            .sorted { $0.part < $1.part }
            .map { (url: dir.appendingPathComponent($0.file), durationSec: $0.durationSec) }
            .filter { exists($0.url) }
        return Selection(urls: kept.map(\.url), durations: kept.map(\.durationSec))
    }

    /// docs/03 ADR-017: the local parts are a seven-day cache now, so a gap in [select]'s playlist
    /// is not necessarily a gap in the recording — Drive keeps every part it was given, and the
    /// detail fetches back what the sweep took.
    ///
    /// The trip is only made when there is a gap *and* Drive has the parts to fill it with: a
    /// recording that never reached Drive has nothing there to ask for, and "No audio on this
    /// device" is still the true sentence about it.
    ///
    /// - Parameters:
    ///   - local: how many parts of the played track this device still has.
    ///   - track: how many the played track has in `meta.json`.
    ///   - uploaded: whether Drive holds every part (`core.uploaded`).
    public static func fetchesFromDrive(local: Int, track: Int, uploaded: Bool) -> Bool {
        local < track && uploaded
    }

    /// The same playlist, rebuilt out of what `core.audio` came back with. A fetched part is
    /// written into the recording's own directory under the name its row gives it, so the
    /// durations the clock counts in are still `meta.json`'s.
    ///
    /// A fetch that could not bring every part back stops at the gap: the playlist is the parts
    /// from the first one up to the one that stayed missing, and nothing after it. Playing on past
    /// a gap would put every later part on the clock at the wrong time — the transcript below reads
    /// against that clock, so 10:00 would index some other moment of the recording. Keeping the
    /// timeline instead (each part played at its `startOffsetSec`) would mean silence between two
    /// queue items, which `AVQueuePlayer` has nothing to make out of; the contiguous prefix is what
    /// can be played correctly, and the detail says in words that the rest could not be fetched.
    ///
    /// - Parameters:
    ///   - parts: the played track's parts, for the order and the durations.
    ///   - files: the file names `core.audio` handed back.
    public static func fetched(parts: [Part_], files: [String], dir: URL) -> Selection {
        let present = Set(files)
        let kept = parts
            .sorted { $0.part < $1.part }
            .prefix { present.contains($0.file) }
        return Selection(
            urls: kept.map { dir.appendingPathComponent($0.file) },
            durations: kept.map(\.durationSec)
        )
    }
}

/// The detail's playback: the parts of one track, end to end, as the one thing the recording was.
///
/// `AVQueuePlayer` is what makes that one thing out of several files — it advances itself, with no
/// gap to hear between two parts of the same take. Several files is also what makes [seek] more
/// than one line: a second of the recording is a second of one particular part, and the queue may
/// have to be rebuilt around it.
@MainActor
public final class RecordingPlayer: ObservableObject {
    @Published public private(set) var isPlaying = false
    /// Seconds from the start of the *recording*, not of the part being played.
    @Published public private(set) var positionSec: Double = 0

    /// How often the playhead is moved while playing: 30 steps a second, so the bar slides rather
    /// than steps. Finer would be redraws no screen this runs on can show.
    public static let tickSec: Double = 1.0 / 30

    private var selection = RecordingPlaylist.Selection.empty
    private var queue: Queue?

    public init() {}

    /// A different recording (or none): whatever was playing stops first, and the queue is the new
    /// one. Building it is left to [play], so picking a row does not touch the audio session.
    ///
    /// Called again at the press of Play, because [stop] leaves the player holding nothing: the
    /// selection that plays is always the one the bar is showing at that moment.
    public func load(_ selection: RecordingPlaylist.Selection) {
        guard selection != self.selection else { return }
        stop()
        self.selection = selection
    }

    public func play() {
        guard !selection.isEmpty else { return }
        (queue ?? makeQueue()).play()
        isPlaying = true
    }

    public func pause() {
        queue?.pause()
        isPlaying = false
    }

    /// Everything that ends playback ends here: the end of the last part, another recording picked,
    /// the surface closed or going away. Idempotent — a player with nothing going has nothing to
    /// stop, and calling it twice is calling it once.
    ///
    /// The next [play] builds the queue again rather than seeking a drained one, which is also what
    /// makes the end of the recording a return to its start.
    ///
    /// What it was going to play goes too. The Mac keeps one player behind a detail view whose
    /// model is swapped per pick, and the new model's bar can offer Play before its own audio is
    /// settled — a player still holding the last recording's selection would answer that press with
    /// the wrong recording. Nothing plays that the caller has not just [load]ed.
    public func stop() {
        queue?.tearDown()
        queue = nil
        selection = .empty
        positionSec = 0
        isPlaying = false
    }

    /// docs/09 화면 원칙 2: a drag on the waveform, or a step of the adjustable action behind it.
    /// The second is the *recording's*, so the first thing it is turned into is a part and an
    /// offset into it ([target]).
    ///
    /// Three cases, and the difference between them is only how much of the queue has to change. A
    /// queue that already has the target part current is seeked where it stands. One that is
    /// somewhere else in the recording is given the items from the target part onwards, because a
    /// queue player cannot be sent backwards through the items it has already drained — the player
    /// itself is kept, so the audio session is neither handed back nor taken again ([Queue.reload]).
    /// And a player that has never been started (or was [stop]ped) gets its queue built here
    /// without being played — pressing Play after a scrub starts from where the scrub left it,
    /// not from zero.
    public func seek(toSec sec: Double) {
        guard !selection.isEmpty else { return }
        let sec = min(max(0, sec), selection.totalSec)
        let target = Self.target(durations: selection.durations, sec: sec)
        // Ahead of the next tick, which does not come at all while the player is paused: the bar
        // is showing the second the finger let go of.
        positionSec = sec
        guard let queue else {
            makeQueue(from: target.index).seek(toSec: target.offsetSec)
            return
        }
        if selection.urls.count - queue.remaining != target.index {
            queue.reload(urls: urls(from: target.index))
            queue.seek(toSec: target.offsetSec)
            // Emptying the queue drops the rate with it, so a player that was going has to be
            // started again — the session it is already holding makes that no more than a `play()`.
            if isPlaying { queue.play() }
            return
        }
        queue.seek(toSec: target.offsetSec)
    }

    /// Where the recording's own clock is: the parts already played, plus how far into the current
    /// one the player is. Pure, because it is the one piece of this worth checking without a file.
    nonisolated static func position(durations: [Double], finished: Int, itemSec: Double) -> Double {
        let played = durations.prefix(max(0, finished)).reduce(0, +)
        return played + (itemSec.isFinite ? max(0, itemSec) : 0)
    }

    /// The other half of [position], and its inverse: which part a second of the recording is in,
    /// and how far into that part. The end of a part is the next part's 0 rather than that part's
    /// last instant, so a scrub to a boundary plays on instead of stopping there. Past the end
    /// clamps to the end of the last part, which is where [seek]'s own clamp already puts it.
    nonisolated static func target(durations: [Double], sec: Double) -> (index: Int, offsetSec: Double) {
        guard let last = durations.indices.last else { return (0, 0) }
        var offsetSec = max(0, sec)
        for (index, durationSec) in durations.enumerated() {
            if offsetSec < durationSec { return (index, offsetSec) }
            offsetSec -= durationSec
        }
        return (last, durations[last])
    }

    /// The parts from [index] to the end of the recording — what a queue holds after a seek into
    /// the middle of it.
    private func urls(from index: Int) -> [URL] {
        Array(selection.urls[min(max(0, index), selection.urls.count - 1)...])
    }

    /// From [index] onwards rather than always from the start, so [seek] can put the queue at a
    /// part the current one has already gone past.
    ///
    /// [tick]'s arithmetic needs nothing said about that: what is left in a queue built from a
    /// suffix is what is left of the *recording* too, so `urls.count - remaining` is still the
    /// number of parts finished.
    private func makeQueue(from index: Int = 0) -> Queue {
        let queue = Queue(
            urls: urls(from: index),
            tick: { [weak self] itemSec, remaining in
                MainActor.assumeIsolated { self?.tick(itemSec: itemSec, remaining: remaining) }
            },
            ended: { [weak self] in
                MainActor.assumeIsolated { self?.stop() }
            }
        )
        self.queue = queue
        return queue
    }

    private func tick(itemSec: Double, remaining: Int) {
        // What is left in the queue is the current item and the ones after it.
        positionSec = Self.position(
            durations: selection.durations,
            finished: selection.urls.count - remaining,
            itemSec: itemSec
        )
    }
}

/// The AVFoundation half of [RecordingPlayer] — the player, the two observers, and the audio
/// session it borrowed — in an object that is *not* on the main actor.
///
/// That is the point of it: `deinit` is nonisolated and cannot call a `@MainActor` method, so a
/// player that goes away without a `stop()` (a window closed while a part is playing) would
/// otherwise leave a periodic observer, a notification registration and an active session behind.
/// Here the last release tears all three down wherever it happens.
private final class Queue {
    private let player: AVQueuePlayer
    private let ended: () -> Void
    private var ticker: Any?
    private var end: NSObjectProtocol?
    #if os(iOS)
    /// Whether *this* player is the one that put the session into playback, and so the one to hand
    /// it back. See [activateSession].
    private var borrowedSession = false
    #endif

    init(urls: [URL], tick: @escaping (Double, Int) -> Void, ended: @escaping () -> Void) {
        let items = urls.map { AVPlayerItem(url: $0) }
        let player = AVQueuePlayer(items: items)
        player.actionAtItemEnd = .advance
        self.player = player
        self.ended = ended
        // A frame's worth: the clock beside the bar only counts whole seconds, but the playhead on
        // the waveform is drawn at this position, and at four steps a second it jumps rather than
        // moves (docs/09 "모션": what moves, moves).
        ticker = player.addPeriodicTimeObserver(
            forInterval: CMTime(seconds: RecordingPlayer.tickSec, preferredTimescale: 600),
            queue: .main
        ) { [weak player] time in
            tick(time.seconds, player?.items().count ?? 0)
        }
        observeEnd(of: items.last)
    }

    deinit { tearDown() }

    /// The current item and the ones after it — what [RecordingPlayer.tick] counts the finished
    /// parts from, asked for outside a tick because [RecordingPlayer.seek] needs it at the press.
    var remaining: Int { player.items().count }

    /// The same player at another part of the recording (see [RecordingPlayer.seek]), because a
    /// queue player cannot be sent back to an item it has already drained.
    ///
    /// The *items* and nothing else: tearing this down and building another would hand the audio
    /// session back with `notifyOthersOnDeactivation` and take it again a moment later, which is
    /// heard as whatever else was playing coming back for that instant. The ticker goes on
    /// reporting through it, so the clock does not blink either.
    func reload(urls: [URL]) {
        player.removeAllItems()
        let items = urls.map { AVPlayerItem(url: $0) }
        for item in items { player.insert(item, after: nil) }
        observeEnd(of: items.last)
    }

    /// Seconds into the *current item*, not into the recording. Both tolerances are zero because
    /// the transcript below is indexed against this clock: landing on the nearest keyframe would
    /// put the playhead somewhere other than where the finger was.
    func seek(toSec sec: Double) {
        player.seek(
            to: CMTime(seconds: sec, preferredTimescale: 600),
            toleranceBefore: .zero,
            toleranceAfter: .zero
        )
    }

    func play() {
        activateSession()
        player.play()
    }

    func pause() {
        player.pause()
        releaseSession()
    }

    /// The queue drains as it plays, so the end of the whole recording is the end of its last part
    /// — after which `AVQueuePlayer` has nothing current and would report no time at all. The last
    /// part is a different item after a [reload], and only one of them is ever watched.
    private func observeEnd(of item: AVPlayerItem?) {
        if let end { NotificationCenter.default.removeObserver(end) }
        end = item.map { last in
            NotificationCenter.default.addObserver(
                forName: .AVPlayerItemDidPlayToEndTime,
                object: last,
                queue: .main
            ) { [ended] _ in ended() }
        }
    }

    /// Idempotent, and safe from `deinit`: nothing here is main-actor state.
    func tearDown() {
        if let ticker { player.removeTimeObserver(ticker) }
        ticker = nil
        if let end { NotificationCenter.default.removeObserver(end) }
        end = nil
        player.pause()
        releaseSession()
    }

    #if os(iOS)
    /// docs/09: playback is the point of pressing Play, so the session says so — without a category
    /// of `.playback` the phone plays nothing while it is on silent.
    ///
    /// Except when the recorder already owns the session. `.playAndRecord` is what a recording in
    /// progress runs in, and playback is allowed inside it — but *setting* a category of `.playback`
    /// on top of it takes the input away and ends the recording. So the category is only ever
    /// changed when it is something else, and the session is only ever handed back by the player
    /// that took it ([borrowedSession]), never by one that merely played inside someone else's.
    private func activateSession() {
        let session = AVAudioSession.sharedInstance()
        guard session.category != .playAndRecord else { return }
        if session.category != .playback {
            try? session.setCategory(.playback)
        }
        try? session.setActive(true)
        borrowedSession = true
    }

    /// Handed back as soon as there is nothing to play, so audio the user paused for this can come
    /// back on its own. Handed back only if it is still ours: a recording that started while this
    /// was playing (a Shortcut, the watch) has since taken the session as `.playAndRecord`, and
    /// deactivating it then would cut that recording off — so the category is checked again here,
    /// not only the flag.
    private func releaseSession() {
        guard borrowedSession else { return }
        borrowedSession = false
        let session = AVAudioSession.sharedInstance()
        guard session.category == .playback else { return }
        try? session.setActive(false, options: .notifyOthersOnDeactivation)
    }
    #else
    private func activateSession() {}
    private func releaseSession() {}
    #endif
}

File: android/app/src/main/kotlin/app/recly/android/ui/RecordingPlayer.kt
package app.recly.android.ui

import android.content.Context
import android.net.Uri
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableDoubleStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.media3.common.AudioAttributes
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.Player
import androidx.media3.common.util.UnstableApi
import androidx.media3.exoplayer.ExoPlayer
import java.io.File

/**
 * The detail's playback: the parts of one track, end to end, as the one thing the recording was.
 *
 * An ExoPlayer playlist is what makes that one thing out of several files — it advances itself,
 * with no gap to hear between two parts of the same take, and it is also what makes a [seek] one
 * call however many parts the second asked for is across. What a caller has of it is "playing or
 * not" and how far in it is. RecKit's `RecordingPlayer` is the same object over `AVQueuePlayer`.
 *
 * The clock is not self-winding: [tick] is called by whoever draws it, four times a second while
 * something is playing (`RecordingDetailScreen`). A player nobody is looking at is not one whose
 * clock anybody needs.
 */
@Stable
@androidx.annotation.OptIn(UnstableApi::class)
class RecordingPlayer(context: Context) {

    var isPlaying by mutableStateOf(false)
        private set

    /** Seconds from the start of the *recording*, not of the part being played. */
    var positionSec by mutableDoubleStateOf(0.0)
        private set

    private var selection = RecordingPlaylist.Selection.EMPTY

    private val player: ExoPlayer = ExoPlayer.Builder(context).build().apply {
        // docs/09: speech, played back — not the recorder's own capture and not a notification.
        setAudioAttributes(
            AudioAttributes.Builder()
                .setContentType(C.AUDIO_CONTENT_TYPE_SPEECH)
                .setUsage(C.USAGE_MEDIA)
                .build(),
            /* handleAudioFocus = */ true,
        )
        repeatMode = Player.REPEAT_MODE_OFF
    }

    // Registered here rather than in the `apply` above, where `stop()` would be the *player's* and
    // not this object's — a call that leaves the recording sitting at its own end with a Pause on
    // the bar, which is what it did until it was moved out.
    init {
        player.addListener(
            object : Player.Listener {
                // The end of the whole recording is the end of its last part, and there is nothing
                // after it: the clock goes back to the start rather than standing at the end.
                override fun onPlaybackStateChanged(playbackState: Int) {
                    if (playbackState == Player.STATE_ENDED) stop()
                }
            },
        )
    }

    /**
     * A different playlist (or none): whatever was playing stops first. Building the queue is left
     * to [play], so opening a recording does not touch the audio output.
     *
     * Called again at the press of Play, because [stop] leaves the player holding nothing: the
     * selection that plays is always the one the bar is showing at that moment.
     */
    fun load(selection: RecordingPlaylist.Selection) {
        if (selection == this.selection) return
        stop()
        this.selection = selection
    }

    fun play() {
        if (selection.isEmpty) return
        queue()
        player.play()
        isPlaying = true
    }

    /**
     * docs/09 화면 원칙 2: the second the finger let go of on the waveform, wherever in the
     * recording it falls. One `seekTo` even when it is in another part, because the playlist is one
     * thing to the player — [RecordingPlaylist.position] and [target] are the same arithmetic read
     * the two ways round.
     *
     * A seek is not a press of Play: a queue is built for a player that is holding nothing, so
     * there is somewhere for the playhead to be, but nothing starts. The clock is moved here rather
     * than left to [tick], which only runs while something is playing.
     */
    fun seek(selection: RecordingPlaylist.Selection, sec: Double) {
        if (selection.isEmpty) return
        val at = sec.coerceIn(0.0, selection.totalSec)
        load(selection)
        queue()
        val (index, offsetSec) = target(selection.durations, at)
        player.seekTo(index, (offsetSec * 1000).toLong())
        positionSec = at
    }

    /** The parts as the player's own playlist, where it is not already holding them. */
    private fun queue() {
        if (player.mediaItemCount > 0) return
        player.setMediaItems(selection.paths.map { MediaItem.fromUri(Uri.fromFile(File(it.toString()))) })
        player.prepare()
    }

    fun pause() {
        player.pause()
        isPlaying = false
    }

    /**
     * Everything that ends playback ends here: the end of the last part, another recording opened,
     * the page left. Idempotent — a player with nothing going has nothing to stop.
     *
     * The next [play] builds the queue again rather than seeking a drained one, which is also what
     * makes the end of the recording a return to its start. What it was going to play goes too, so
     * nothing plays that the caller has not just [load]ed.
     */
    fun stop() {
        player.stop()
        player.clearMediaItems()
        selection = RecordingPlaylist.Selection.EMPTY
        positionSec = 0.0
        isPlaying = false
    }

    /**
     * The recording's own clock, as of now — the parts already played plus the current one.
     *
     * Only while something is playing: the caller's ticking coroutine is cancelled a frame or two
     * after [stop], and a stray tick in between would put the clock back where the recording ended
     * instead of at its start.
     */
    fun tick() {
        if (!isPlaying) return
        positionSec = RecordingPlaylist.position(
            durations = selection.durations,
            // The current item is the one after all the finished ones.
            finished = player.currentMediaItemIndex,
            itemSec = player.currentPosition / 1000.0,
        )
    }

    /** The player is gone after this. Nothing else may be called on it. */
    fun release() {
        player.release()
        isPlaying = false
    }

    companion object {
        /**
         * How often the playhead is moved while playing: 30 steps a second, so the bar slides rather
         * than steps — the clock beside it only counts whole seconds, but the playhead on the
         * waveform is drawn at this position, and at four steps a second it jumps rather than moves
         * (docs/09 "모션"). Finer would be redraws no screen this runs on can show.
         */
        const val TICK_MS: Long = 33

        /**
         * Which part a second of the recording is in, and how far into it — the item and the offset
         * a [seek] is made of. Pure, because it is the whole of what a scrub across a part boundary
         * has to get right; the Windows shell's `RecordingPlayer.target` is the same arithmetic.
         *
         * The boundary belongs to the part that starts there rather than to the one that ends
         * there, so a drag through it lands on the first frame of the next part instead of the last
         * of the previous. Past the end is the end of the last part, and no parts at all is the
         * start of a playlist that has nothing in it to be anywhere else in.
         */
        fun target(durations: List<Double>, sec: Double): Pair<Int, Double> {
            val last = durations.lastIndex
            if (last < 0) return 0 to 0.0
            var offsetSec = sec.coerceAtLeast(0.0)
            durations.forEachIndexed { index, durationSec ->
                if (offsetSec < durationSec) return index to offsetSec
                offsetSec -= durationSec
            }
            return last to durations[last]
        }
    }
}
